testsuite/tests/effects/dynamic.ml

```diff
@@ -0,0 +1,254 @@
+                    go (); Thread.join t;
+                    Printf.printf "with_temporarily still [expect %d]: %d\n" n (Dynamic.get d)));
+   Printf.printf "after with_temporarily [expect %d]: %d\n" outside (Dynamic.get d))
+
+let _ = let d = Dynamic.make 42 in
+  (Dynamic.set_root d 7; test_with_temp d 6)
+
+(* In the presence of effects, check that set_root values are visible in other fibers,
+   without contaminating the value on a pre-existing thread. *)
+
+let _ =
+
+  let n = 10 in
+  let outside = n+10 in
+  let d = Dynamic.make outside in
+  let check_other_thread, finish =
+    sync_thread (fun () -> Printf.printf "In pre-existing thread [expect %d]: %d\n" outside (Dynamic.get d)) in
+
+  let f () =
+    (Printf.printf "In fiber [expect %d]: %d\n" n (Dynamic.get d);
+     check_other_thread();
+     Dynamic.set_root d (n+1);
+     Effect.perform (E ());
+     Printf.printf "In continuation [expect %d]: %d\n" (n+2) (Dynamic.get d);
+     Dynamic.set_root d (n+3)) in
+
+  let h : type a. a Effect.t -> ((a, 'b) Effect.Deep.continuation -> 'b) option = function
+    | E () -> Some (fun k ->
+      Printf.printf "in handler [expect %d]: %d\n" (n+1) (Dynamic.get d);
+      check_other_thread();
+      Dynamic.set_root d (n+2);
+      Effect.Deep.continue k ();
+      Printf.printf "after continuation [expect %d]: %d\n" (n+4) (Dynamic.get d))
+    | e -> None in
+
+  Dynamic.set_root d n;
+  Effect.Deep.match_with f ()
+    { retc = (fun () -> (Printf.printf "after fiber [expect %d]: %d\n" (n+3) (Dynamic.get d);
+                         check_other_thread();
+                         Dynamic.set_root d (n+4)));
+      exnc = (fun e -> raise e);
+      effc = h };
+  finish()
+
+(* Does with_temporarily work correctly in effect handlers? *)
+
+let _ =
+
+  let n = 20 in
+  let outside = n+10 in
+  let d = Dynamic.make outside in
+  let check_other_thread, finish =
+    sync_thread (fun () -> Printf.printf "In pre-existing thread [expect %d]: %d\n" outside (Dynamic.get d)) in
+
+  let f () =
+    (Printf.printf "In fiber [expect %d]: %d\n" n (Dynamic.get d);
+     check_other_thread();
+     Dynamic.set_root d (n+1);
+     Effect.perform (E ());
+     Printf.printf "In continuation [expect %d]: %d\n" (n+3) (Dynamic.get d);
+     Dynamic.set_root d (n+4)) in
+
+  let h : type a. a Effect.t -> ((a, 'b) Effect.Deep.continuation -> 'b) option = function
+    | E () -> Some (fun k ->
+      Printf.printf "in handler [expect %d]: %d\n" (n+1) (Dynamic.get d);
+      test_with_temp d (n+2);
+      check_other_thread();
+      Dynamic.set_root d (n+3);
+      Effect.Deep.continue k ();
+      Printf.printf "after continuation [expect %d]: %d\n" (n+5) (Dynamic.get d))
+    | e -> None in
+
+  (Dynamic.set_root d n;
+   Effect.Deep.match_with f ()
+     { retc = (fun () -> (Printf.printf "after fiber [expect %d]: %d\n" (n+4) (Dynamic.get d);
+                          check_other_thread();
```

Comment:
Don't think it'd be any different, but we could also check `with_temporarily` works in `retc`/`exnc`